File: src/execution/runtimeQueueEngine.ts
/**
 * Converted from Python: core/execution/runtime_queue_engine.py
 * @generated — WebWeaveX python→javascript library port
 */

import * as py from "../runtime/pyCompat.js";

export let MAX_QUEUE: any = 100000;
export function enqueueRuntimeAction(queue: any, action: any, priority: any = 0): any {
  var updated: any = [...py.iter(queue)];
  var entry: any = {"action": action, "priority": priority, "order": py.len(updated)};
  py.listAppend(updated, entry);
  updated = py.slice(py.sorted(updated, {key: ((item: any) => [(-py.toInt(py.get(item, "priority", 0))), py.toInt(py.get(item, "order", 0))]) as (item: any) => any}), null, MAX_QUEUE);
  return {"queue": updated, "size": py.len(updated), "bounded": true};
}
export function dequeueRuntimeAction(queue: any): any {
  if (!py.truthy(queue)) {
    return {"queue": [], "action": null, "bounded": true};
  }
  var ordered: any = py.sorted(queue, {key: ((item: any) => [(-py.toInt(py.get(item, "priority", 0))), py.toInt(py.get(item, "order", 0))]) as (item: any) => any});
  var head: any = py.at(ordered, 0);
  var rest: any = py.slice(ordered, 1, null);
  return {"queue": rest, "action": py.get(head, "action"), "bounded": true};
}

File: lib/src/execution/runtime_queue_engine.dart
/// Port of core/execution/runtime_queue_engine.
const int maxQueue = 100000;

int _asInt(dynamic v, int fallback) {
  if (v is int) return v;
  if (v is num) return v.toInt();
  if (v is String) return int.tryParse(v) ?? fallback;
  return fallback;
}

int _queueCompare(Map<String, dynamic> a, Map<String, dynamic> b) {
  // key = (-priority, order)
  final int pa = -_asInt(a['priority'], 0);
  final int pb = -_asInt(b['priority'], 0);
  final int c = pa.compareTo(pb);
  if (c != 0) return c;
  return _asInt(a['order'], 0).compareTo(_asInt(b['order'], 0));
}

Map<String, dynamic> enqueueRuntimeAction(
  List<Map<String, dynamic>> queue,
  Map<String, dynamic> action, {
  int priority = 0,
}) {
  final List<Map<String, dynamic>> updated = <Map<String, dynamic>>[...queue];
  updated.add(<String, dynamic>{
    'action': action,
    'priority': priority,
    'order': updated.length,
  });
  final List<Map<String, dynamic>> sorted = <Map<String, dynamic>>[...updated]
    ..sort(_queueCompare);
  final List<Map<String, dynamic>> bounded =
      sorted.length > maxQueue ? sorted.sublist(0, maxQueue) : sorted;
  return <String, dynamic>{
    'queue': bounded,
    'size': bounded.length,
    'bounded': true,
  };
}

Map<String, dynamic> dequeueRuntimeAction(List<Map<String, dynamic>> queue) {
  if (queue.isEmpty) {
    return <String, dynamic>{
      'queue': <Map<String, dynamic>>[],
      'action': null,
      'bounded': true,
    };
  }
  final List<Map<String, dynamic>> ordered = <Map<String, dynamic>>[...queue]
    ..sort(_queueCompare);
  final Map<String, dynamic> head = ordered.first;
  final List<Map<String, dynamic>> rest = ordered.sublist(1);
  return <String, dynamic>{
    'queue': rest,
    'action': head['action'],
    'bounded': true,
  };
}
